---
title: "Analysing Vicky's run of RNA Interactome Capture data for Tunicamycin/ISRIB treated U20S cells"
author: "Manasa Ramakrishna"
date: "31/08/2018"
output: html_notebook
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=12,fig.height=8,warning=FALSE, message=FALSE)
#tidy.opts=list(width.cutoff=80)
```

```{r A_Startup, hide=T,warning=FALSE, message=FALSE}
#---------------------------------------------------------------------------
# Author 	      : Manasa Ramakrishna, mr325@le.ac.uk
# Date started 	: 9th January, 2018
# Last modified : 9th January, 2018
# Aim 		      : To take a look at first TMT-labelled MCF10A data on insulin starved and unstarved samples
# Depends       : On 'tmtFunctions.R'. Make sure they are in the same directory
# Notes         : Works on data from Veronica's latest TMT-labelled proteomics data
#--------------------------------------------------------------------------- 

# Invoking libraries
library(MSnbase)
library(gplots)
library(reshape2)
library(wesanderson)
library(sva)
library(splitstackshape)
#library(GenomicFeatures)
library(mygene)
library(ggplot2)
#library(devtools)
#install_github("vqv/ggbiplot")
library(ggbiplot)
library(limma)
library(RColorBrewer)
library(calibrate)
library(VennDiagram)
#devtools::install_github("guiastrennec/ggplus")
library(data.table)
library(goseq)
library(ggplus)
library(pRoloc)
library(pRolocdata)
library(pRolocGUI)


#Setting working directories
wd = "/Users/manasa/Documents/Work/TTT/02_Proteomics/13_Vicky-U20S-RIC-Tunicamycin/"
setwd(wd)
getwd()

indir = paste(wd,"Input",sep="/")
outdir = paste(wd,paste(Sys.Date(),"Output",sep = "_"),sep = "/")

source("../12_Veronica-RIC-insulin/mcf10aFunctions.R")
source("../12_Veronica-RIC-insulin/tmtFunctions.R")

if (exists(outdir)){
  print("Outdir exists")
}else{
  dir.create(outdir)
}

```

```{r 01: RNA-binding-proteins-from-Synapt}
#------------------------------------------------------------------------------------------------------------------------
# Step 01: Adding in data from Synapt based, label-free mass spectrometry data from Leicester
#------------------------------------------------------------------------------------------------------------------------

dat = read.table(paste(indir,"tunicamycin-ric.txt",sep="/"),sep="\t",header=T,stringsAsFactors = F)
print(dim(dat))
s = apply(dat[,3:10],2,summary)
print(s)
print(summary(rowSums(dat[,3:10] == 0)))

samples = data.frame(cbind(Source = rep("RIC",8), Samples = colnames(dat)[3:10],RBPs = as.numeric(as.character(nrow(dat))),Missing = colSums(dat[,3:10] == 0)))
samples$RBPs = as.numeric(as.character(samples$RBPs))
samples$Missing = as.numeric(as.character(samples$Missing))
samples$RBPs = samples$RBPs-samples$Missing


# Melting the data for a barplot
ds.melt = melt(samples, id=c("Source", "Samples"))
ds.melt <- within(ds.melt, Samples <- factor(Samples, levels=c("DMSO_A","DMSO_B","TUNIC_A","TUNIC_B","TUNICISRIB_A","TUNICISRIB_B","ISRIB_A","ISRIB_B")))
ds.melt$value = ds.melt$value*100/612

# Drawing a barplot of the relative proportions of Missing and RBPs using DIA and DDA with various callers
# Missing means "NA" so no abundance values and no peptides for the protein.

pdf(paste(outdir,paste(format(Sys.time(),"%Y-%m-%d_%H.%M.%S"),"Barplot-showing-proportion-of-RBPs-across-conditions-callers.pdf",sep="_"),sep="/"),paper="a4r",width=12,height=8)
p <-ggplot(ds.melt)+ geom_bar(aes(Samples, value,fill=variable),stat = "identity", width = 0.5)+labs(y = "Percentage.of.proteins")+theme_bw()+
scale_fill_manual("legend", values = c("Missing" = "#541F12", "RBPs" = "#CC8B3C"))+theme(text = element_text(size=15), plot.title = element_text(hjust = 0.5,vjust=0.5),axis.ticks.x = element_blank())+ggtitle("Missing values across the experiments")
print(p)
dev.off()
```

```{r 02a_RBPs-filter}

#==================================================================================
# 02a : Filtering RBPs in the data set for most promising candidates
#==================================================================================

sc.dat = dat
sc.dat$nas = rowSums(sc.dat == 0)
rownames(sc.dat) = dat$Accession.Number
table(sc.dat$nas)
dim(sc.dat)

#  0   1   2   3   4   5   6   7 
#168  62  45  50  51  49  76 111 

#----------------------------------------------------------------------------------------------------------------------------------
# Filt 1 : First filter is to remove any proteins who are missing more than 6 values out of 8. This excludes genes where not even 
# a single experimental conditions has a complete set of abundance calls. 
#----------------------------------------------------------------------------------------------------------------------------------
filt.1 = sc.dat[which(sc.dat$nas <= 6),]
dim(filt.1) # 501 proteins

#----------------------------------------------------------------------------
# Look into normalising data usng ALBU_BOVIN/ADH1_YEAST as DMSO_B seems high
#----------------------------------------------------------------------------

norm1 = norm2 = norm3 = norm4 = data.frame()
for(i in 1:nrow(filt.1)){
  norm1 = rbind(norm1,filt.1[i,3:10]/filt.1[1,3:10]) # Normalise using ALBU
  norm2 = rbind(norm2,filt.1[i,3:10]/filt.1[2,3:10]) # Normalise using ALDH1
  norm3 = rbind(norm3,filt.1[i,3:10]/filt.1[3,3:10]) # Normalise using TRYP
  norm4 = rbind(norm3,filt.1[i,3:10]/(colSums(filt.1[1:3,3:10])/3)) # Normliase using ALBU/ALDH1/TRYP
}

# More normalisation
norm5 = filt.1[,3:10]/colMeans(filt.1[,3:10]) # Normalise using column means
norm6 = filt.1[,3:10]/rowMeans(filt.1[,3:10]) # Normalise using row means

# Some heatmaps to compare normalisation strategies
#--------------------------------------------------
mypalette <- brewer.pal(11,"RdYlBu")
morecols <- colorRampPalette(mypalette)

pdf(paste(outdir,paste(format(Sys.time(),"%Y-%m-%d_%H.%M.%S"),"Heatmaps-showing-various-normalisations.pdf",sep="_"),sep="/"),paper="a4r",width=12,height=8)
heatmap.2(as.matrix(filt.1[,3:10]),col=rev(morecols(50)),trace="none",main="Expression of proteins using SCAFFOLD + No normalisation",scale="row",labRow = rownames(filt.1) ,keysize=1,key.title = "",tracecol="black",cexRow = 0.3,dendrogram="column", srtCol=45, cexCol = 1)
heatmap.2(as.matrix(norm1),col=rev(morecols(50)),trace="none",main="Expression of proteins using SCAFFOLD + ALBU_BOVIN Normalisation",scale="row",labRow = rownames(filt.1) ,keysize=1,key.title = "",tracecol="black",cexRow = 0.3,dendrogram="column", srtCol=45, cexCol = 1)
heatmap.2(as.matrix(norm2),col=rev(morecols(50)),trace="none",main="Expression of proteins using SCAFFOLD + ADH1_YEAST Normalisation",scale="row",labRow = rownames(filt.1) ,keysize=1,key.title = "",tracecol="black",cexRow = 0.3,dendrogram="column", srtCol=45, cexCol = 1)
heatmap.2(as.matrix(norm3),col=rev(morecols(50)),trace="none",main="Expression of proteins using SCAFFOLD + TRYP_PIG Normalisation",scale="row",labRow = rownames(filt.1) ,keysize=1,key.title = "",tracecol="black",cexRow = 0.3,dendrogram="column", srtCol=45, cexCol = 1)
heatmap.2(as.matrix(norm4),col=rev(morecols(50)),trace="none",main="Expression of proteins using SCAFFOLD + Column Means Normalisation",scale="row",labRow = rownames(filt.1) ,keysize=1,key.title = "",tracecol="black",cexRow = 0.3,dendrogram="column", srtCol=45, cexCol = 1)
heatmap.2(as.matrix(norm5),col=rev(morecols(50)),trace="none",main="Expression of proteins using SCAFFOLD + Row Means Normalisation",scale="row",labRow = rownames(filt.1) ,keysize=1,key.title = "",tracecol="black",cexRow = 0.3,dendrogram="column", srtCol=45, cexCol = 1)
heatmap.2(as.matrix(norm6),col=rev(morecols(50)),trace="none",main="Expression of proteins using SCAFFOLD + Row Means Normalisation",scale="row",labRow = rownames(filt.1) ,keysize=1,key.title = "",tracecol="black",cexRow = 0.3,dendrogram="column", srtCol=45, cexCol = 1)
dev.off()

# Boxplots to compare normalisation strategies
#----------------------------------------------------------------------------
pdf(paste(outdir,paste(format(Sys.time(),"%Y-%m-%d_%H.%M.%S"),"Boxplots-showing-various-normalisations.pdf",sep="_"),sep="/"),paper="a4r",width=12,height=8)
par(mfrow=c(3,2))
boxplot(filt.1[,3:10], cex.axis = 0.7, main="No Norm")
boxplot(norm1, cex.axis = 0.7, main = "ALBUMIN Norm")
boxplot(norm2, cex.axis = 0.7, main = "ALDH1 Norm")
#boxplot(norm3, cex = 0.8, main = "TRYP Norm")
boxplot(norm4, cex.axis = 0.7, main = "ALBU-ALDH1-TRYP Norm")
boxplot(norm5, cex.axis = 0.7, main = "Column means Norm")
boxplot(norm6, cex.axis = 0.7, main = "Row means Norm")
dev.off()

# PCA plots to compare normalisation strategies
#----------------------------------------------------------------------------
pdf(paste(outdir,paste(format(Sys.time(),"%Y-%m-%d_%H.%M.%S"),"PCA-plots-showing-various-normalisations.pdf",sep="_"),sep="/"),paper="a4r",width=12,height=8)
names = c("No Norm","ALBUMIN Norm","ALDH1 Norm","TRYP Norm","ALBU-ALDH1-TRYP Norm","Column means Norm","Row means Norm")
i = 1
for(k in list(filt.1[,3:10],norm1,norm2,norm3,norm4,norm5,norm6)){
  pal = wes_palette("Darjeeling1")
  samples = rep(c("DMSO","TUNIC","TUNICISRIB","ISRIB"),each=2)
  pca = prcomp(t(k))
  summary(pca)
  p <- ggbiplot(pca,choices=c(1,2), var.axes=F, groups = samples, circle = T,labels=rownames(pca$x),colour = pal[as.factor(samples)])+labs(title = paste("PCA plot for ", "RIC-Scaffold with ", names[i],sep=""))+scale_colour_manual(name = "Condition",values=c("DMSO" = "#00A08A","TUNIC" = "#F98400","TUNICISRIB" = "#F2AD00","ISRIB" = "#FF0000"))+   theme(plot.title = element_text(hjust = 0.5))
  print(p)
  i = i+1
}
dev.off()

```
Having looked at various normalisation strategies, the division of all rows by the expression of their respective ALBU_BOVIN works best in clustering DMSO_A and DMSO_B together and away from the others. Will use this for downstream filtering and analysis. 

```{r 2b_RBPs-filter-continued}

#----------------------------------------------------------------------------------------------------------------------------------
# Filt 2 : Then remove common mass spec contaminants and spikeins - keratins, bovine albumin based on CCP list of contaminants
#----------------------------------------------------------------------------------------------------------------------------------
contam = read.table("Input/contam.txt",sep="\t",header=F,stringsAsFactors = F)
filt.2 = norm1[-which(rownames(norm1) %in% contam$V1),]
dim(filt.2) # 487 proteins, only 14 contaminants 

#----------------------------------------------------------------------------------------------------------------------------------
# Filt 3 : Replace missing values with 0. Ading a column called "zeroes" so I can tell how many were 0 and how many were NAs
# This is not really a filter but I need to do this to be able to draw a PCA
#----------------------------------------------------------------------------------------------------------------------------------
filt.3 = filt.2
filt.3$zeroes = rowSums(filt.3 == 0)
filt.3[is.na(filt.3)] = 0
dim(filt.3) # 487 proteins

# Filt4 is for expression cut-offs but since normalisation reduces them to fractions, so I've lef this out. 

#----------------------------------------------------------------------------------------------------------------------------------
# Filt 5: Just keep the columns with necessary information
#----------------------------------------------------------------------------------------------------------------------------------
filt.5 = filt.3[,1:8]
dim(filt.5) # 487 proteins
```

```{r 03_Plotting-filtered-data}

#==================================================================================
# 03 : Taking a closer look at filtered data
#==================================================================================

# Some stats and plots for each caller
pal = wes_palette("Darjeeling1")
samples = rep(c("DMSO","TUNIC","TUNICISRIB","ISRIB"),each=2)

#----------------------------------
# PCA for each caller
#----------------------------------
pdf(paste(outdir,paste(format(Sys.time(),"%Y-%m-%d_%H.%M.%S"),"PCA-plots-1-2-across-samples.pdf",sep="_"),sep="/"),paper="a4r",width=12,height=8)
pca = prcomp(t(filt.5))
summary(pca)
p <- ggbiplot(pca,choices=c(1,2), var.axes=F, groups = samples, circle = T,labels=rownames(pca$x),colour = pal[as.factor(samples)])+labs(title = paste("PCA plot for ", "RIC-Scaffold",sep=""))+scale_colour_manual(name = "Condition",values=c("DMSO" = "#00A08A","TUNIC" = "#F98400","TUNICISRIB" = "#F2AD00","ISRIB" = "#FF0000"))+ theme(plot.title = element_text(hjust = 0.5))
print(p)
dev.off()

#----------------------------------
# Draw heatmap of data by caller
# Can change to better colours
#----------------------------------
pdf(paste(outdir,paste(format(Sys.time(),"%Y-%m-%d_%H.%M.%S"),"Heatmaps-of-filtered-data-by-caller.pdf",sep="_"),sep="/"),paper="a4r",width=12,height=8)
mypalette <- brewer.pal(11,"RdYlBu")
morecols <- colorRampPalette(mypalette)
heatmap.2(as.matrix(filt.5),col=rev(morecols(50)),trace="none",main="Expression of proteins called based on DIA using SCAFFOLD",scale="row",labRow = rownames(filt.5) ,keysize=1,key.title = "",tracecol="black",cexRow = 0.3,dendrogram="column", srtCol=45, cexCol = 1)
dev.off()


```

```{r 04_Pulling-out-most-interesting-proteins}

#-------------------------------------------------------------------------------------------------------------------------------------
# Pulling out proteins common to all callers to look at their trend in protein expression (based on venn diagram 'v')
#-------------------------------------------------------------------------------------------------------------------------------------
common.prots = filt.5[order(rownames(filt.5)),]
common.prots$ID = rownames(common.prots)
write.table(common.prots$ID,paste(outdir,"Filtered-protein-list.txt",sep="/"),sep="\t",row.names=F,quote=F)

# Melt the dataframe for ggplot compatibility
# Keep all conditions for a given gene in order
com.melt = melt(common.prots,id="ID")
colnames(com.melt)[2:3] = c("Condition","Exp")
com.melt$Exp = as.numeric(com.melt$Exp)
com.melt$Tmt = sapply(strsplit(as.character(com.melt$Condition),"\\_"),"[[",1)
com.melt$Tmt = factor(com.melt$Tmt,levels = c('DMSO','TUNIC','TUNICISRIB','ISRIB'),ordered = TRUE)
com.melt = com.melt[order(com.melt$ID),]

# Excludes Isoquant as this is the only absolute quantification and units are in fmol. Rest are spectral counts
pdf(paste(outdir,paste(format(Sys.time(),"%Y-%m-%d_%H.%M.%S"),"Filtered-common-genes-across-callers-violinplots.pdf",sep="_"),sep="/"),paper="a4r",width=12,height=8)
gg2 = ggplot(data=com.melt,aes(x=Tmt, y=Exp, group = Tmt,fill=Tmt)) + geom_violin() +stat_summary(aes(group=1),fun.y=median, geom="line", color="black", size=0.5,lty = 2) + theme(legend.position="top",axis.text.x = element_text(angle = 60, hjust = 1))+labs(fill = "Treatment")+ scale_fill_manual(values=c("#E2D200","#46ACC8","#E58601","#B40F20"))+facet_wrap(~ID)
gg10 <- facet_multiple(plot=gg2, facets="ID", ncol = 4, nrow = 4, scales = "free_y")
dev.off()
```

```{r 05a_A-simple-classification}
# Objective is to classify each of the proteins depending on their trends in expression across conditions
# 4 conditions - DMSO, ISRIB, TUNIC, ISRIBTUNIC and can be "Up","Down". 2^4 = 16 combinations

filt.6 = cbind(rowMeans(filt.5[,1:2]),rowMeans(filt.5[,3:4]),rowMeans(filt.5[,5:6]),rowMeans(filt.5[,7:8]),"")
colnames(filt.6) = colnames=c("DMSO","TUNIC","TUNICISRIB","ISRIB","Class")
filt.6 = as.data.frame(filt.6,stringsAsFactors = F)
filt.6[,1:4] = apply(filt.6[,1:4],2,function(x) as.numeric(x))

# Add class to each protein
for (p in 1:nrow(filt.6)){
  status = ""
  a
  # Is the trend from DMSO to Tunicamycin up or down ?
  if(filt.6[p,"DMSO"]<filt.6[p,"TUNIC"]){
    status = paste(status,"Up-",sep="")
  }
  else{
    status = paste(status,"Down-",sep="")
  }
  
  # Is the trend from Tunicamycin to Tunicamycin+Isrib up or down ?
  if(filt.6[p,"TUNIC"]<filt.6[p,"TUNICISRIB"]){
    status = paste(status,"Up-",sep="")
  }
  else{
    status = paste(status,"Down-",sep="")
  }  
  
  # Is the trend from Tunicamycin+Isrib to just Isrib up or down ?
  if(filt.6[p,"TUNICISRIB"]<filt.6[p,"ISRIB"]){
    status = paste(status,"Up",sep="")
  }
  else{
    status = paste(status,"Down",sep="")
  }
  #print(status)
  filt.6$Class[p] = status
}

# Now plot the trends
filt.6$Names = rownames(filt.6)
melt6 = melt(filt.6, id.vars = c("Names","Class"))
lgp = ggplot(melt6,aes(x=variable,y=value))+geom_line(aes(col=Class,group=Names))+facet_wrap(~Class,scales = "free_y")

pdf(paste(outdir,paste(format(Sys.time(),"%Y-%m-%d_%H.%M.%S"),"Patterns-of-protein-expression-across-conditions.pdf",sep="_"),sep="/"),paper="a4r",width=12,height=8)
print(lgp)
dev.off()

```
```{r 05b_A-secondary-classification}
# Objective is to classify each of the proteins depending on their trends in expression across conditions
# 4 conditions - DMSO, ISRIB, TUNIC, ISRIBTUNIC and can be "Up","Down". 2^4 = 16 combinations

filt.7 = filt.6
filt.7$Class = ""

# Add class to each protein
for (p in 1:nrow(filt.7)){
  status = ""
  # Is the trend from DMSO to Tunicamycin up or down ?
  if(filt.7[p,"DMSO"]<filt.7[p,"TUNIC"]){
    status = paste(status,"Up-",sep="")
  }
  else{
    status = paste(status,"Down-",sep="")
  }
  
  # Is the trend from Tunicamycin to Tunicamycin+Isrib up or down ?
  if(filt.7[p,"DMSO"]<filt.7[p,"TUNICISRIB"]){
    status = paste(status,"Up-",sep="")
  }
  else{
    status = paste(status,"Down-",sep="")
  }  
  
  # Is the trend from Tunicamycin+Isrib to just Isrib up or down ?
  if(filt.7[p,"DMSO"]<filt.7[p,"ISRIB"]){
    status = paste(status,"Up",sep="")
  }
  else{
    status = paste(status,"Down",sep="")
  }
  #print(status)
  filt.7$Class[p] = status
}

# Now plot the trends
filt.7$Names = rownames(filt.7)
melt7 = melt(filt.7, id.vars = c("Names","Class"))
lgp = ggplot(melt7,aes(x=variable,y=value))+geom_line(aes(col=Class,group=Names))+facet_wrap(~Class,scales = "free_y")

pdf(paste(outdir,paste(format(Sys.time(),"%Y-%m-%d_%H.%M.%S"),"Patterns-of-protein-expression-across-conditions.pdf",sep="_"),sep="/"),paper="a4r",width=12,height=8)
print(lgp)
dev.off()
```


```{r 06a_Functional-enrichment}

#---------------------------------------------------------------------------------------------------------------
# Functional enrichment of proteins using whole proteome as background
#---------------------------------------------------------------------------------------------------------------

# Reading in whole proteome ids form uniprot to obtain protein length to provide as bias to goseq
mcf10prot = read.delim("../12_Veronica-RIC-insulin/Input/Whole-proteome-all-id-mapping.tab",sep="\t",header=T)
dim(mcf10prot)

upcom = read.delim("Input/tunicamycin-ric-with-uniprot-ids.tab",sep="\t",header=T)
rownames(upcom) = upcom$Entry

prots.ann = merge(filt.6,upcom, by.x = "Names", by.y = "Entry.name",all.x=T)
colnames(prots.ann)[7] = "Entry.name"
prots.ann$Genes = sapply(strsplit(as.character(prots.ann$Gene.names),"\\s+"),"[[",1)

# GO analysis
univ = mcf10prot$Entry
univ.ann = readRDS("Input/MCF10A-mapped-proteins-kegg.rds")
univ.ann$kegg.id = sapply(univ.ann$pathway.kegg, function(x) paste0(unique(unlist(x[[1]])),collapse=";"))
univ.ann$kegg.name = sapply(univ.ann$pathway.kegg, function(x) paste0(unique(unlist(x[[2]])),collapse=";"))

# Make mapping for goseq analysis
univ.cat.go = makeGene2Cat(univ.ann,"query","go.all",";") 
univ.cat.doms = makeGene2Cat(univ.ann,"query","domains",";")
univ.cat.kegg = makeGene2Cat(univ.ann,"query","kegg.name",";")

# Running functional enrichment
library(goseq)
all.go = NULL
all.pro = NULL
all.kegg = NULL

for(l in unique(prots.ann$Class)){
  prots = as.character(prots.ann$Entry[which(prots.ann$Class == l)])
  go = rungoseq(prots,univ.cat.go,univ, b=mcf10prot$Length, 0.05)
  pro = rungoseq(prots,univ.cat.doms,univ, b=mcf10prot$Length, 0.05)
  kegg = rungoseq(prots,univ.cat.kegg,univ, b=mcf10prot$Length, 0.05)
  
  # Save results  
  if(nrow(go[[2]])>0){
    all.go = rbind(all.go,cbind(Cluster=l,go[[2]]))
  }
  if(nrow(pro[[2]])>0){
    all.pro = rbind(all.pro,cbind(Cluster=l,pro[[2]]))
  }
  if(nrow(kegg[[2]])>0){
    all.kegg = rbind(all.kegg,cbind(Cluster=l,kegg[[2]]))
  }
}

# Plot enrichments ?
all.go$Description = paste("(",all.go$ontology,") ",all.go$term,sep="")
all.pro$Description = all.pro$category
all.kegg$Description = all.kegg$category

colnames(all.kegg)[3] = colnames(all.pro)[3] = colnames(all.go)[3] = "pvalue"

# Plots
ego = enricherPlot(all.go,"All-GO-Tunic-Isrib",N=15,colorBy = "neg.log10.BH",sizeBy ="foldEnrich",low.col="#E69F00",high.col="#999999",trunc.len=40,all.size=10,y.size=12,x.size=9)
epro = enricherPlot(all.pro,"All-Interpro-Tunic-Isrib",N=15,colorBy = "neg.log10.BH",sizeBy ="foldEnrich",low.col="#E69F00",high.col="#999999",trunc.len=40,all.size=10,y.size=12,x.size=9)
ekegg = enricherPlot(all.kegg,"All-KEGG-Tunic-Isrib",N=15,colorBy = "neg.log10.BH",sizeBy ="foldEnrich",low.col="#E69F00",high.col="#999999",trunc.len=40,all.size=10,y.size=12,x.size=9)

pdf(paste(outdir,paste(format(Sys.time(),"%Y-%m-%d_%H.%M.%S"),"Go-Interpro-KEGG-enrichment-plots.pdf",sep="_"),sep="/"),paper="a4r",width=12,height=8)
print(ego)
print(epro)
print(ekegg)
dev.off()

# Re-order all.go and all.pro
all.go[1:5,c(1:2,12,15,13,16,3,8:9,20,5:6,10:11,17)]
all.go.1 = all.go[,c(1:2,12,15,13,16,3,8:9,20,5:6,10:11,17)]
all.pro[1:5,c(1:2,10,13,11,14,3,5:6,8:9,15)]
all.pro.1 = all.pro[,c(1:2,10,13,11,14,3,5:6,8:9,15)]
all.kegg[1:5,c(1:2,10,13,11,14,3,5:6,8:9,15)]
all.kegg.1 = all.kegg[,c(1:2,10,13,11,14,3,5:6,8:9,15)]

write.table(all.go.1,paste(outdir,"All-Enriched-GO-terms.txt",sep="/"),sep="\t",row.names=F,quote=F)
write.table(all.pro.1,paste(outdir,"All-Enriched-Interpro-domains.txt",sep="/"),sep="\t",row.names=F,quote=F)
write.table(all.kegg.1,paste(outdir,"All-Enriched-KEGG-pathways.txt",sep="/"),sep="\t",row.names=F,quote=F)

```


```{r}


comprots.ann = merge(common.prots,upcom, by.x = "ID", by.y = "Entry.name",all.x=T)
colnames(comprots.ann)[1] = "Uniprot.name"

# Checking the overlap with whole proteome. 38/362 are not present in the whole proteome
# It seems as if MATR3 is misisng from the whole proteome but it just is under a different Uniprot ID which is "A8MXP9" instead of P43243
miss.wholprot = setdiff(upcom$Entry,mcf10prot$Entry)
upcom[intersect(upcom$Entry,miss.wholprot),]

# GO analysis
univ = mcf10prot$Entry
#univ.ann = myProtMapper(univ,out.fields=c("interpro.short_desc","ensembl.gene","pathway.kegg.name","pathway.kegg.id","go.MF.id","go.CC.id","go.BP.id"))
#saveRDS(univ.ann,"Input/MCF10A-mapped-proteins-kegg.rds")
univ.ann = readRDS("Input/MCF10A-mapped-proteins-kegg.rds")
univ.ann$kegg.id = sapply(univ.ann$pathway.kegg, function(x) paste0(unique(unlist(x[[1]])),collapse=";"))
univ.ann$kegg.name = sapply(univ.ann$pathway.kegg, function(x) paste0(unique(unlist(x[[2]])),collapse=";"))

# Make mapping for goseq analysis
univ.cat.go = makeGene2Cat(univ.ann,"query","go.all",";") 
univ.cat.doms = makeGene2Cat(univ.ann,"query","domains",";")
univ.cat.kegg = makeGene2Cat(univ.ann,"query","kegg.name",";")

#Run goseq
library(goseq)
go = rungoseq(comprots.ann$Entry,univ.cat.go,univ, b=mcf10prot$Length, 0.05)
pro = rungoseq(comprots.ann$Entry,univ.cat.doms,univ,b=mcf10prot$Length, 0.05)
kegg = rungoseq(comprots.ann$Entry,univ.cat.kegg,univ,b=mcf10prot$Length, 0.05)
  
write.table(go[[2]][,c(1,6:8,4:5,9:10,14,11:12,15)],paste(outdir,"All-proteins-GO-enrichment.txt",sep="/"),sep="\t",quote=F,row.names=F)
write.table(pro[[2]][,c(1,4:8,12,9:10,13)],paste(outdir,"All-proteins-Interpro-enrichment.txt",sep="/"),sep="\t",quote=F,row.names=F)
write.table(pro[[2]][,c(1,4:8,12,9:10,13)],paste(outdir,"All-proteins-Interpro-enrichment.txt",sep="/"),sep="\t",quote=F,row.names=F)

```

```{r 06_DE-analysis}

samp.dat = colnames(filt.5)
samp.dat = cbind(samp.dat,Tag=paste("TMT",1:8,sep=""))
colnames(samp.dat)[1] = "Sample"
samp.dat = data.frame(samp.dat,stringsAsFactors = F)
samp.dat$Rep = rep(c(1,2),2)
rownames(samp.dat) = samp.dat$Sample
samp.dat$Sample = sapply(strsplit(samp.dat$Sample,"\\_"),"[[",1)

colnames(comprots.ann)[11] = "Uniprot.name"
rownames(comprots.ann) = comprots.ann$Uniprot.name
filt.5 = filt.5[order(rownames(filt.5)),]

# Need to have rownames for both fData and expression matrix in the same order. 
res <- MSnSet(exprs = as.matrix(filt.5),fData=comprots.ann[,c(1,11:16)],pData = samp.dat)
fData(res)$gene = sapply(strsplit(rownames(fData(res)),"_"),"[[",1)

# Design a model matrix
des = model.matrix(~0+pData(res)$Sample)
colnames(des) = gsub("pData\\(res\\)\\$Sample","",colnames(des))
rownames(des) = rownames(pData(res))

# Draw contrasts for groups of interest
contr <- makeContrasts(
  isrib.vs.dmso = ISRIB - DMSO,
  tunic.vs.dmso = TUNIC - DMSO, 
  tunic.isrib.vs.dmso = TUNICISRIB - DMSO,
  tunic.vs.isrib = TUNIC - ISRIB,
  tunicisrib.vs.isrib = TUNICISRIB - ISRIB,
  tunicisrib.vs.tunic = TUNICISRIB - TUNIC,
  levels = des)
contr

# Set up results directory
res.dir = paste(outdir,paste(format(Sys.time(),"%Y-%m-%d_%H.%M.%S"),"All-results",sep="_"),sep="/")

# Create the results directory to make it easier to look through results
if (exists(res.dir)){
  print("Results directory exists")
}else{
  dir.create(res.dir)
}

# Running DE for all comparisons
limma.cond = runDE(res,des,contr,res.dir,logfc=0.5,pval=0.05)

for(c in colnames(contr)){
  plotDE(limma.cond[[2]],res,logfc.col="logFC",cpm.col= NULL,"P.Value","adj.P.Val",pval=0.05,logfc=1,suf=c,out.dir=res.dir,cont.col.names=gsub(".vs.","|",c))
}


# Not many DE genes so checking expression state
t = apply(exprs(res),2,table)
dt = plyr::ldply(t, rbind)
dt[is.na(dt)] = 0

rownames(dt) = dt$.id
dt = dt[,2:ncol(dt)]
dt = t(dt)

# Scaling data
sc.dat = dat[,3:10]/colSums(dat[,3:10])
```

```{r Cluster-Extras}

# Could try some kmeans clustering
library(factoextra)
res.dist <- get_dist(filt.5, stand = TRUE, method = "pearson")
fviz_dist(res.dist,gradient = list(low = "#00AFBB", mid = "white", high = "#FC4E07"))

par(mfrow=c(2,2))
fviz_nbclust(filt.5, cluster::clara, "silhouette")
fviz_nbclust(filt.5, cluster::clara, "wss")
fviz_nbclust(filt.5, cluster::clara, "gap_stat")

# 
km.res <- kmeans(filt.5, 3, nstart = 25)

# Visualize
fviz_cluster(km.res, data = filt.5,
             ellipse.type = "convex",
             palette = "jco",
             ggtheme = theme_minimal())

par(mfrow=c(2,2))
for(i in 1:3){
  n = names(km.res$cluster[which(km.res$cluster==i)])
  temp = filt.5[which(rownames(filt.5) %in% n),]
  matplot(t(temp),type="l",col=rainbow(4)[i],cex = 0.8,main=paste("Cluster number : ",i,sep=""), ylab = "Normalised abundance",xlab = "Samples: 1-2 = DMSO; 3-4 = TUNIC; 5-6 = TUNICISRIB; 7-8 = ISRIB")
}

library(NbClust)
nb <- NbClust(filt.5, distance = "euclidean", min.nc = 2,max.nc = 10, method = "kmeans")

```